new project YandexSamokat autotest UI with POM

Zorkiy87 opened the pull request.

Zorkiy87 committed:
new project YandexSamokat autotest UI with POM

ShamilovR commented on src/pages/base_page.py:
```diff
@@ -0,0 +1,56 @@
+
+    @allure.step('Нажатие на кнопку Принять Куки')
+    def click_cookie(self):
+        self.click_element(BasePageLocators.ACCEPT_COOKIE)
```
Нужно исправить: базовый класс это абстракция. Он не должен работать с конкретными элементами, локаторами и урлами.

ShamilovR commented on tests/test_main_page.py:
```diff
@@ -0,0 +1,97 @@
+    @allure.title('Проверка ответа на вопрос: сколько стоит?')
+    def test_check_text_important_questions_how_much(self, driver):
+        main_page = MainPage(driver)
+        locator = QuestionsLocators.HOW_MUCH
```
Нужно исправить: здесь и далее: в классическом POM тест не должен знать о внутренней реализации, в том числе и о используемых локаторах (за исключением передачи локаторов в рамках параметризации). Подобная реализация должна быть сокрыта в рамках методов пейджи.

ShamilovR commented on src/pages/main_page.py:
```diff
@@ -0,0 +1,31 @@
+        questions_page.click_element(locator)
+        questions_page.wait_for_element_visible(question)
+        description = questions_page.get_description(question)
+        assert description == answer
```
Нужно исправить: это логика самого тестового метода. ее не нужно прятать в пейджи.

ShamilovR commented on src/pages/main_page.py:
```diff
@@ -0,0 +1,31 @@
+
+    @allure.step('Проверка ответов на основные вопросы на главной странице')
+    def test_check_text_important_questions (self, driver, locator, question, answer):
+        questions_page = MainPage(driver)
```
Нужно исправить: внутри класса не нужно создать экземпляр этого же класса. Для этого используется self

ShamilovR commented on src/pages/main_page.py:
```diff
@@ -0,0 +1,31 @@
+class MainPage(BasePage):
+
+    @allure.step('Проверка ответов на основные вопросы на главной странице')
+    def test_check_text_important_questions (self, driver, locator, question, answer):
```
Нужно исправить: вся работа с драйвером должна быть изолирована в базовом классе. Дочернии пейджи не должны работать с инстансом драйвера

ShamilovR commented on tests/test_main_page.py:
```diff
@@ -0,0 +1,97 @@
+import allure
+import src.config
+
+from conftest import driver
```
Нужно исправить: здесь и далее: фикстуры не нужно импортировать. Если они лежат в conftest то pytest их видит без импорта

Zorkiy87 replied:
Почему-то не видит фикстуры из conftest если они в корне проекта. Перенёс в папку test, заработало.

Zorkiy87 committed:
1) Из базового класса убрал методы работы с конкретными элементами;
2) Из test_main_page перенёс в main_page саму работу тестов;
3) Перёнес из main_page.py в test_main_page логику проверки тестов на соответствие вопроса ответу;
4) Поправил, сейчас дочерние пейджи не работают с инстансом драйвера;
5) Убрал импорт фикстур;
6) Доработал название тестов и параметров, для лучшего понимания кода.

ShamilovR commented on tests/test_main_page.py:
```diff
@@ -0,0 +1,93 @@
+        actual_answer = main_page.get_description(answer_locator)
+        assert actual_answer == expected_answer, f"Ожидался ответ: '{expected_answer}', но получен '{actual_answer}'"
+
+class TestImportantQuestions(ImportantQuestionsTestBase):
```
Нужно исправить: тестовые классы используются только для объединения тестовых методов в сьюты. Они не должны использовать наследование

ShamilovR commented on tests/test_main_page.py:
```diff
@@ -0,0 +1,93 @@
+        main_page.click_element(question_locator)
+        main_page.wait_for_element_visible(answer_locator)
+        actual_answer = main_page.get_description(answer_locator)
+        assert actual_answer == expected_answer, f"Ожидался ответ: '{expected_answer}', но получен '{actual_answer}'"
```
Нужно исправить: здесь и далее: тестовые модули должны содержать только тестовые классы и тестовые сценарии.

Zorkiy87 committed:
1) Убрал наследование в test_page
2) В test_page оставил только сами тесты. verify_question_answer вынес в main_page
